### ISSUE TITLE
More general padding

### ISSUE THREAD

**Issue by ebetica** [pytorch/pytorch#1825]
It would be useful to have a padding function like numpy's pad. I think for PyTorch, what makes sense would be a `functional.constant_pad(value, [(dim_0_before, dim_0_after), (dim_1_before, dim_1_after), ...])`, where each tuple can also be a single element for before and after. Or, something like `F.constant_pad(value, (before, after), dim=x)`. I suggest we have separate functions for each type of padding (e.g. F.edge_pad, F.reflection_pad) instead of numpy's monolithic [pad](https://docs.scipy.org/doc/numpy-1.12.0/reference/generated/numpy.pad.html) function.

**Comment by apaszke** [comment]
http://pytorch.org/docs/nn.html?highlight=pad#torch.nn.functional.pad

**Comment by ebetica** [comment]
For example, I want to pad a 3-D tensor (time, x, y) along the time dimension. The existing pad does not allow me to do this. @apaszke

**Comment by apaszke** [comment]
Right. I think we can support both fixed-size tuples (and assume that input has `len(tuple) + 2` dims), or accept a list of paddings for all dims

**Comment by soumith** [comment]
now an Nd padding is merged into master via #2657 and it supports padding All dimensions (not just the last N-2 dimensions.
